src/lib/lockfile.ts

```diff
@@ -1,34 +1,44 @@
-import { promises as fs, unlinkSync, rmdirSync } from 'fs';
+import { promises as fs } from 'fs';
+import type { Stats, Dirent } from 'fs';
 import os from 'os';
 import { dirname } from 'path';
 
 import { exec } from './fs-utils';
+import { isENOENT, isEISDIR, isEPERM } from './errors';
 
 // Equivalent to `drwxrwxrwt`
 const STICKY_WRITE_PERMISSIONS = 0o1777;
 
-/**
- * Internal lockfile manager to track files in memory
- */
-// Track locksTaken, so that the proper locks can be cleaned up on process exit
-const locksTaken: { [lockName: string]: boolean } = {};
-
-// Returns all current locks taken, as they've been stored in-memory.
+// Returns all current locks taken under a directory (default: /tmp)
 // Optionally accepts filter function for only getting locks that match a condition.
-export const getLocksTaken = (
-	lockFilter: (path: string) => boolean = () => true,
-): string[] => Object.keys(locksTaken).filter(lockFilter);
-
-// Try to clean up any existing locks when the process exits
-process.on('exit', () => {
-	for (const lockName of getLocksTaken()) {
-		try {
-			unlockSync(lockName);
-		} catch (e) {
-			// Ignore unlocking errors
+// A file is counted as a lock by default if it ends with `.lock`.
+export const getLocksTaken = async (
```

Comment:
I don't think we need to query the state of the directory each time, which is a bit wasteful. I think maybe this module could have something like an `initialize` function that receives the initial state of locks and then when new locks are taken and released the state gets updated there. Then `getLocksTaken` remains the same.

On the other hand, the `update-lock` module, also can have an initialization method that reads the directories once (that is called by application manager).

Reply:
Added back the in-memory store in https://github.com/balena-os/balena-supervisor/pull/2234/commits/d8d211cf7610008ac5754d35654e8f3f06856e05, skipped some lock tests in actions as a result, as it's not possible to manipulate the in-memory lockfile store of balena-supervisor-sut from the `sut` test process

Reply:
I added a comment on one of those tests. While you cannot easily fake the in memory lock store, locks taken by the user app should still prevent certain requests from taking place.

Reply:
Also, just for the sake of argument. Assuming the user app is behaving properly (i.e. not overwriting our locks), can you think of a case where the in-memory store can get out of sync with the filesystem? I want to make sure we are covering all of our bases

Reply:
We discussed this last checkin (Mar 28 2024) and agreed to remove the in-memory store in case lockfile interference causes the store to get out-of-sync with the actual lock state in the filesystem.